feat: Manage Team ooo or crud

vijayraghav-io opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #16333 
- Fixes CAL-

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

Approach to fetch past records:-  Infinite scrolling with pagination is implemented to fetch additional records (10 at a time) when the user scrolls to the bottom of the table. 
Also search by username and email is implemented to fetch past records of a member quickly.

https://www.loom.com/share/fe88b4739846419493fe26836daedf88?sid=96acca39-eaad-4fcc-89e9-c2153532e080

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] - N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
feature for team ooo crud and past records

anikdhabal commented on packages/features/settings/OutOfOfficeEntriesListForTeam.tsx:
```diff
@@ -0,0 +1,280 @@
+import { Avatar, Button, DataTable, EmptyScreen, Icon, showToast, SkeletonText, Tooltip } from "@calcom/ui";
+import type { BookingRedirectForm } from "@calcom/web/pages/settings/my-account/out-of-office";
+
+interface OutOfOfficeEntry {
+  id: number;
+  uuid: string;
+  start: Date;
+  end: Date;
+  toUserId: number | null;
+  toUser: {
+    username: string;
+  } | null;
+  reason: {
+    id: number;
+    emoji: string;
+    reason: string;
+    userId: number;
+  } | null;
+  notes: string | null;
+  user: { id: number; avatarUrl: string; username: string; email: string } | null;
+}
```
This is used twice, can you pls reuse it another component?

vijayraghav-io replied:
yeah, it can be reused, will update

anikdhabal commented on packages/features/settings/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -25,17 +26,34 @@ export const CreateOrEditOutOfOfficeEntryModal = ({
+    value: number;
+    label: string;
+  }[] =
+    listMembers
+      ?.filter((member) => (oooType === "mine" ? me?.data?.id !== member.id : oooType === "team"))
```
It’s better to use enum instead of hardcoded values like "mine" and "team."

vijayraghav-io replied:
ok

Amit91848 commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -68,9 +115,22 @@ const OutOfOfficePage = () => {
             setCurrentlyEditingOutOfOfficeEntry(null);
           }}
           currentlyEditingOutOfOfficeEntry={currentlyEditingOutOfOfficeEntry}
+          oooType={oooType}
+          setTeamOOOEntriesUpdated={setTeamOOOEntriesUpdated}
+          setMyOOOEntriesUpdated={setMyOOOEntriesUpdated}
+        />
+      )}
+      {oooType === "team" ? (
+        <OutOfOfficeEntriesListForTeam
+          editOutOfOfficeEntry={editOutOfOfficeEntry}
+          teamOOOEntriesUpdated={teamOOOEntriesUpdated}
+        />
+      ) : (
+        <OutOfOfficeEntriesList
+          editOutOfOfficeEntry={editOutOfOfficeEntry}
+          myOOOEntriesUpdated={myOOOEntriesUpdated}
```
Why are there separate components for Team and Mine view?

As far as I can tell both are same with same components and code?

vijayraghav-io replied:
they are similar, but differ in columns displayed. Wanted to avoid having many conditions within same file for team and mine and keep it separated , s

vijayraghav-io replied:
handled in single component

anikdhabal commented on packages/features/settings/OutOfOfficeEntriesListForTeam.tsx:
```diff
@@ -0,0 +1,280 @@
+      },
+      {
+        getNextPageParam: (lastPage) => lastPage.nextCursor,
+        placeholderData: keepPreviousData,
+      }
+    );
+
+  useEffect(() => {
+    refetch();
+  }, [teamOOOEntriesUpdated, deletedEntry, refetch]);
+
+  const totalDBRowCount = data?.pages?.[0]?.meta?.totalRowCount ?? 0;
+  //Flatten the array of arrays from the useInfiniteQuery hook
+  const flatData = useMemo(
+    () => data?.pages?.flatMap((page) => page.rows) ?? [],
+    [data]
+  ) as OutOfOfficeEntry[];
+  const totalFetched = flatData.length;
+
+  const columns: ColumnDef<OutOfOfficeEntry>[] = [
+    {
+      id: "member",
+      header: `Member`,
+      cell: ({ row }) => {
+        if (!row.original || !row.original.user) {
+          return <SkeletonText className="h-8 w-full" />;
+        }
+        const { avatarUrl, username, email } = row.original.user;
+        return (
+          <div className="flex items-center gap-2">
+            <Avatar
+              size="sm"
+              alt={username || email}
+              imageSrc={getUserAvatarUrl({
+                avatarUrl,
+              })}
+            />
+            <div className="">
+              <div
+                data-testid={`member-${username}-username`}
```
This is similar to the `OutOfOfficeEntriesList`. No need to create another component, instead use it accordingly.

Amit91848 commented:
Unable to scroll Redirect To Member when editing an entry, can only scroll it while adding a new that too I have to close it and open it for it to work, can you also replicate it?


https://github.com/user-attachments/assets/103278f8-dc90-4cb5-b058-93274247d4fd

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -366,8 +535,35 @@ export const outOfOfficeEntriesList = async ({ ctx }: { ctx: { user: NonNullable
+        : false,
+    },
+    orderBy: {
+      end: "desc",
```
Imo it is better to sort by start date rather than end.

vijayraghav-io replied:
updated

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -290,11 +357,32 @@ export const outOfOfficeEntryDelete = async ({ ctx, input }: TBookingRedirectDel
     throw new TRPCError({ code: "BAD_REQUEST", message: "out_of_office_id_required" });
   }
 
+  let deletingOooOfUserId = ctx.user.id;
+  let deletingOooOfEmail = ctx.user.email;
+  let deletingOooOfUserName = ctx.user.username;
```
I don't like the variable names in the whole file; they need to be improved

vijayraghav-io replied:
i was trying to be more descriptive for purpose of new variables, updated, hope it is better now.

Amit91848 reviewed (changes requested):
Great stuff as usual @vijayraghav-io , but this needs some more improvement.

For orgs I can add redirection from one team to another


https://github.com/user-attachments/assets/bbf63c27-5a33-41c0-ab15-9d8ff7f54964

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -335,16 +423,97 @@ export const outOfOfficeEntryDelete = async ({ ctx, input }: TBookingRedirectDel
+    }
+
+    // Fetch team member user ids
+    const members = await prisma.membership.findMany({
+      where: {
+        teamId: {
+          in: teams.map((team) => team.teamId),
+        },
+        userId: {
+          not: ctx.user.id,
+        },
+      },
+      select: {
+        userId: true,
+      },
+    });
+    if (members.length === 0) {
+      throw new TRPCError({ code: "NOT_FOUND", message: "no_team_members" });
+    }
+    fetchOOOEntriesForIds = members.map((member) => member.userId);
+  }
```
This is a very performance-reducing operation. Need to be improved

vijayraghav-io replied:
updated

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -335,16 +423,97 @@ export const outOfOfficeEntryDelete = async ({ ctx, input }: TBookingRedirectDel
-      start: "asc",
+  });
+
+  const outOfOfficeEntries = await prisma.outOfOfficeEntry.findMany({
```
Can you not add _count in select for this, instead of having separate queries for count and actual data?

vijayraghav-io replied:
The first query gets the total count of matching records, while the second query fetches only limited records (say 10). I tried _count can't be used here.

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -4,11 +4,48 @@ import dayjs from "@calcom/dayjs";
 import { TRPCError } from "@trpc/server";
 
-import type { TOutOfOfficeDelete, TOutOfOfficeInputSchema } from "./outOfOffice.schema";
+import type {
+  TOutOfOfficeDelete,
+  TOutOfOfficeEntriesListSchema,
+  TOutOfOfficeInputSchema,
+} from "./outOfOffice.schema";
+
+// function getTeam() checks if there is a team where 'adminUserId' is admin or owner
+// and 'memberUserId' is a member.
+// If exists returns the first found such team.
+const getTeam = async (adminUserId: number, memberUserId: number) => {
+  return await prisma.team.findFirst({
+    where: {
+      AND: [
+        {
+          members: {
+            some: {
+              userId: adminUserId,
+              role: {
+                in: [MembershipRole.ADMIN, MembershipRole.OWNER],
+              },
+            },
+          },
+        },
+        {
+          members: {
+            some: {
+              userId: memberUserId,
+            },
+          },
+        },
+      ],
+    },
```
Too many changes in the backend inside a single file. We definitely need to add tests

anikdhabal commented on packages/features/settings/OutOfOfficeEntriesListForTeam.tsx:
```diff
@@ -0,0 +1,280 @@
+  user: { id: number; avatarUrl: string; username: string; email: string } | null;
+}
+
+export const OutOfOfficeEntriesListForTeam = ({
+  editOutOfOfficeEntry,
+  teamOOOEntriesUpdated,
+}: {
+  editOutOfOfficeEntry: (entry: BookingRedirectForm) => void;
+  teamOOOEntriesUpdated: number;
+}) => {
+  const { t } = useLocale();
+  const tableContainerRef = useRef<HTMLDivElement>(null);
+  const [debouncedSearchTerm, setDebouncedSearchTerm] = useState("");
+  const [deletedEntry, setDeletedEntry] = useState(0);
+  const { data, isPending, fetchNextPage, isFetching, refetch } =
+    trpc.viewer.outOfOfficeEntriesList.useInfiniteQuery(
+      {
+        limit: 10,
+        fetchTeamMembersEntries: true,
+        searchTerm: debouncedSearchTerm,
+      },
+      {
+        getNextPageParam: (lastPage) => lastPage.nextCursor,
+        placeholderData: keepPreviousData,
+      }
+    );
+
+  useEffect(() => {
+    refetch();
+  }, [teamOOOEntriesUpdated, deletedEntry, refetch]);
+
+  const totalDBRowCount = data?.pages?.[0]?.meta?.totalRowCount ?? 0;
+  //Flatten the array of arrays from the useInfiniteQuery hook
+  const flatData = useMemo(
+    () => data?.pages?.flatMap((page) => page.rows) ?? [],
+    [data]
+  ) as OutOfOfficeEntry[];
+  const totalFetched = flatData.length;
+
```
Hey, this is similar to the other component. I know there are a few different cases that don't match, but instead of doing the same things twice in two different components, can you please try to consolidate them into a single component? If things become too difficult and there are too many visible conditional changes, it's okay to define 2 components. However, in such cases, define things once and use them in another component. For example, EmptyScreen is defined twice, etc.

vijayraghav-io replied:
handled in single component

anikdhabal reviewed (changes requested):
Great work @vijayraghav-io , don't test it locally but codewise I've some doubt.

vijayraghav-io commented:
Thankyou! @Amit91848 @anikdhabal , sure will address the comments.

vijayraghav-io committed:
added tests for ooo handler, combined list components for mine,team

vijayraghav-io commented:
> Unable to scroll Redirect To Member when editing an entry, can only scroll it while adding a new that too I have to close it and open it for it to work, can you also replicate it?

I checked multiple times, not able to recreate this for me. Able to search the list of members for redirect also.

vijayraghav-io committed:
nit: removed commented code

CarinaWolli reviewed (changes requested):
I didn't review the code but tested it locally, found the following issues: 

- [ ] We need rounded corners in both my and team OOO:
![Screenshot 2024-09-05 at 10 47 37 AM](https://github.com/user-attachments/assets/fbadc18e-b7e8-448f-b3f5-2936ac4ef3d8)
- [ ] Non-admins/owners should not see the Team OOO tab at all 
- [ ] Can we show past OOO in different tab called 'previous'? 
		As described here: https://github.com/calcom/cal.com/issues/16333#issuecomment-2307482194

vijayraghav-io committed:
added current-previous tab , not to show TeamOOO tab for nonadmin

vijayraghav-io commented:
@CarinaWolli , Thanks for review 🙏 

1. added current-previous tab
My view:
<img width="811" alt="Screenshot 2024-09-06 at 3 52 19 PM" src="https://github.com/user-attachments/assets/be7f340c-8704-4a32-8595-0774686937f6">
Team view : 
<img width="811" alt="Screenshot 2024-09-06 at 4 07 56 PM" src="https://github.com/user-attachments/assets/dd093c97-8a39-4ab7-a4a2-77a130eaff4a">

2. Updated not to show `Team OOO` tab for non-admin/owners

3. For Rounded Corners -
Since the `DataTable` component is generic, its taking more time to update this than i expected. Will try to update this.
Also i have one doubt - should the rounded corners be there for all items in the list, or only for first and last one.

vijayraghav-io commented on packages/features/settings/OutOfOfficeEntriesList.tsx:
```diff
@@ -1,47 +1,263 @@
+  }
+  columns.push({
+    id: "outOfOffice",
+    header: `OutOfOffice (${totalDBRowCount})`,
```
updated

vijayraghav-io committed:
update for type checks

CarinaWolli commented:
> For Rounded Corners -
> Since the DataTable component is generic and any changes here may affect any other views, its taking more time to update this than i expected. Will try to update this.
> Also i have one doubt - should the rounded corners be there for all items in the list, or only for first and last one.

I see, it should be rounded corners on all other tables as well. Like on the team availability view: 
![Screenshot 2024-09-06 at 12 51 49 PM](https://github.com/user-attachments/assets/5b726fd8-ee60-4aa7-a1ee-83906615fb27)
Only the 4 outside corners should be rounded same as in the team member list for example: 
![Screenshot 2024-09-06 at 12 52 32 PM](https://github.com/user-attachments/assets/b86a3d69-c5af-40a4-a105-f52b317b49fd)

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
updated tests after merge

vijayraghav-io commented:
> I see, it should be rounded corners on all other tables as well. Like on the team availability view: 

@CarinaWolli , Thanks for confirming, but i tried with existing classes like `rounded-lg`,`rounded-md`,... and tried with new class like `rounded-[10px]`. Though i can see in browser debugger 10px radius gets applied, for some reason this is not visible.

vijayraghav-io commented:
@Amit91848 @anikdhabal , have addressed your comments.

vijayraghav-io committed:
updated outofoffice handler after merge

vijayraghav-io committed:
for alignment with outofoffice header column

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.test.ts:
```diff
@@ -0,0 +1,1396 @@
+      const result = await outOfOfficeCreateOrUpdate({ ctx, input });
+      expect(result).toStrictEqual({});
+
+      //expect outOfOffice to be created in db for userId:1
+      const ooo = await prismock.outOfOfficeEntry.findMany({
+        where: {
+          userId: users[0].id,
+        },
+      });
+      expect(ooo.length).toEqual(1);
+
+      //expect the start and end dates are stored in db are correct
+      expect(ooo[0].start).toEqual("2024-10-01T00:00:00.000Z");
+      expect(ooo[0].end).toEqual("2024-10-01T23:59:59.999Z");
+    });
+
+    it("should successfully create outofoffice with redirect for logged-in user.", async () => {
+      const teams = [
+        {
+          id: 1,
+          name: "Team 1",
+          slug: "team-1",
+        },
+      ];
+      await addTeamsToDb(teams);
+
+      const users = [
+        {
+          id: 1,
+          name: "User 1",
+          username: "user-1",
+          email: "user-1@test.com",
+          timeZone: "Asia/Kolkata",
+          schedules: [TestData.schedules.IstWorkHours],
+        },
+        {
+          id: 2,
+          name: "User 2",
+          username: "user-2",
+          email: "user-2@test.com",
```
We already have end-to-end tests for this common operation related to OOO entries. Please test the specific changes you made and focus primarily on team OOO cases. The current changes make the test unnecessarily larger. Also, use 1 or 2 items in the list or array of items as examples.

vijayraghav-io replied:
Since we are modifying the same function to create ooo for individual user as well as for team admin, I think it's better to cover both scenarios for unit tests.

I have used minimum possible users for each scenario, for team admin creating ooo scenario - we need atleast three - 
1. admin (context user), 
2. user for whom ooo is being created
3. user for whom ooo is redirected.

In this test, I have used only 3.

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.test.ts:
```diff
@@ -0,0 +1,1396 @@
+      ];
+      await addTeamsToDb(teams);
+
+      const users = [
+        {
+          id: 1,
+          name: "User 1",
+          username: "user-1",
+          email: "user-1@test.com",
+          timeZone: "Asia/Kolkata",
+          schedules: [TestData.schedules.IstWorkHours],
+        },
+        {
+          id: 2,
+          name: "User 2",
+          username: "user-2",
+          email: "user-2@test.com",
+          timeZone: "Asia/Kolkata",
+          schedules: [TestData.schedules.IstWorkHours],
+        },
+        {
+          id: 3,
+          name: "User 3",
+          username: "user-3",
+          email: "user-3@test.com",
+          timeZone: "Asia/Kolkata",
+          schedules: [TestData.schedules.IstWorkHours],
+        },
+      ];
```
As I mentioned, use a short example when a big list of items is not needed.

vijayraghav-io replied:
Please refer reply to previous comment, here as well 3 users are required,
user[0] is admin,
user[1] is user for whom ooo is  created with redirect to user[2]

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
update after resolving merge conflicts

vijayraghav-io committed:
nit/chore: spacing

vijayraghav-io committed:
update after changes in router migration

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
ui improvements rounded corner

vijayraghav-io committed:
updated props

CarinaWolli commented on packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -118,9 +147,43 @@ export const CreateOrEditOutOfOfficeEntryModal = ({
+                  ? t("edit_an_out_of_office")
+                  : oooType === "team"
+                  ? t("create_ooo_dialog_team_title")
+                  : t("create_an_out_of_office")
```
I would add an subtitle to the dialog 'Create an OOO entry for a team member'

And [below](https://github.com/calcom/cal.com/pull/16456/files#diff-eb35f039b483b5448706495223947f00a7a7d7a485ae1f386469c574843e1ea8R162) just use 'Team member' as a label

vijayraghav-io replied:
ok sure, updated , PFB screenshot of updated Dialog UI

<img width="512" alt="Screenshot 2024-09-21 at 2 29 57 PM" src="https://github.com/user-attachments/assets/abea2f73-cf86-4a95-b03e-4ab8c643e53a">

CarinaWolli commented on packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -118,9 +147,43 @@ export const CreateOrEditOutOfOfficeEntryModal = ({
+
+            {/* In case of Team, Select Member for whom OOO is created */}
+            {oooType === "team" && (
+              <div className="mb-4 mt-4 h-16">
```
Can we make sure the spaces are all the same. There is not enough space between label and dropdown and too much space under dropdown: 
![Screenshot 2024-09-20 at 2 17 08 PM](https://github.com/user-attachments/assets/485a1f08-ca09-4bf0-82b6-2e206bd2ff1d)

vijayraghav-io replied:
updated , It was like that for Redirect TeamMember selection before and somehow overlooked this one, updated there as well.
<img width="509" alt="Screenshot 2024-09-21 at 2 34 41 PM" src="https://github.com/user-attachments/assets/950b1063-fe48-44db-a98c-293df8bcad49">

CarinaWolli commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -2603,10 +2603,22 @@
+  "edit_ooo_team_label": "Edit out of office for team member",
+  "create_ooo_dialog_team_title": "Create Out of Office",
+  "ooo_team_empty_title": "Create OOO for your team",
+  "ooo_team_empty_description": "Communicate to your team member's bookers when you're team not available to take bookings. They can still book you're team upon their return or you can forward them to other team member.",
```
```suggestion
  "ooo_team_empty_description": "Communicate to bookers when your team members are not available to take bookings. They can still be booked upon their return or you can forward them to another team member.",
```

CarinaWolli commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -2603,10 +2603,22 @@
+  "create_ooo_dialog_team_title": "Create Out of Office",
+  "ooo_team_empty_title": "Create OOO for your team",
+  "ooo_team_empty_description": "Communicate to your team member's bookers when you're team not available to take bookings. They can still book you're team upon their return or you can forward them to other team member.",
+  "forward_to_team_member_only": "Invalid Redirection!. OOO Forwarding can be done to a member of same team.",
```
```suggestion
  "forward_to_team_member_only": "OOO Forwarding is only allowed to a member of the same team.",
```

CarinaWolli replied:
```suggestion
  "forward_to_team_member_only": "OOO Forwarding is only allowed to a member of the same team.",
```

CarinaWolli commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -63,8 +141,7 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
               members: {
                 some: {
-                  userId: ctx.user.id,
-                  accepted: true,
```
why was `accepted: true `removed?

vijayraghav-io replied:
when i started to develop, i remember i found a valid reason for this and felt its not necessary that's why i had removed, forgot to note down the exact reason.
May be did i think that already in listMembers.handler.ts we are doing, and result from this only we are showing in selection.
 https://github.com/calcom/cal.com/blob/6dd8812a19635456f7e8168aa480e2f16d5eb687/packages/trpc/server/routers/viewer/teams/listMembers.handler.ts#L34
 
Added back, if i recollect a strong reason for removing this, will update. Anyway it should not break in most of the cases.

CarinaWolli reviewed (dismissed):
Found a few month things that need to be addressed first 🙏

vijayraghav-io committed:
Update apps/web/public/static/locales/en/common.json

Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>

vijayraghav-io committed:
Update apps/web/public/static/locales/en/common.json

Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>

vijayraghav-io committed:
updated dialog subtitles, review comments

vijayraghav-io commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -1,26 +1,84 @@
+    data && (data.user.role === MembershipRole.OWNER || data.user.role === MembershipRole.ADMIN);
+  const isOrgAndPrivate = data?.isOrganization && data.isPrivate;
+  const toggleGroupOptions = [{ value: OutOfOfficeTab.MINE, label: t("my_ooo") }];
+  if (isOrgAdminOrOwner && !isOrgAndPrivate) {
```
ok, updated

vijayraghav-io commented:
> Found a few month things that need to be addressed first 🙏

@CarinaWolli Thankyou! 🙏  , i have addressed the review comments.

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -10,12 +10,60 @@ import type { OOOEntryPayloadType } from "@calcom/features/webhooks/lib/sendPayl
+  type TOutOfOfficeInputSchema,
+} from "./outOfOffice.schema";
+
+// function getTeam() checks if there is a team where 'adminUserId' is admin or owner
+// and 'memberUserId' is a member.
+// If exists returns true.
+const getTeam = async (adminUserId: number, memberUserId: number) => {
```
Please rename this function. getTeam() suggestes you are returning some team but it checks if current user has privilege to edit memberUser. You can then also remove the comment

vijayraghav-io replied:
done

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -51,19 +101,47 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
+  let oooUserOrgId = ctx.user.organizationId;
+  let oooUserFullName = ctx.user.name;
+
+  // Check If Admin or Owner is trying to create OOO for their team member, and is valid.
```
NIT: remove comment

vijayraghav-io replied:
done

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -403,11 +485,32 @@ export const outOfOfficeEntryDelete = async ({ ctx, input }: TBookingRedirectDel
+  let oooUserName = ctx.user.username;
+
+  if (input.userId && input.userId !== ctx.user.id) {
+    // Check if context user is admin for the OOOEntry's member
```
NIT: remove comment

vijayraghav-io replied:
removed

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
I think now would be a great time to refactor this file. It has handlers to create/update, delete and list ooo which it too much for a single file. We should have separate handler files for each route

vijayraghav-io replied:
created separate handler files

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.schema.ts:
Same for this schema file. Break it down into separate files

vijayraghav-io replied:
created separate schema files

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -479,8 +664,35 @@ export const outOfOfficeEntriesList = async ({ ctx }: { ctx: { user: NonNullable
+        : false,
+    },
+    orderBy: {
+      start: "desc",
```
```suggestion
      start: "asc",
```

I think this should be ascending, closest ooo should be shown first

vijayraghav-io replied:
ok

Amit91848 reviewed (dismissed):
Superb work as usual @vijayraghav-io. Left some suggestions

Amit91848 committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

Amit91848 commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.schema.ts:
```diff
@@ -16,6 +17,22 @@ export type TOutOfOfficeInputSchema = z.infer<typeof ZOutOfOfficeInputSchema>;
+  cursor: z.number().nullish(),
+  fetchTeamMembersEntries: z.boolean().optional().default(false),
+  searchTerm: z.string().optional(),
+  recordType: z.enum(["current", "previous"]).default(OutOfOfficeRecordType.CURRENT),
```
```suggestion
  recordType: z.nativeEnum(OutOfOfficeRecordType).default(OutOfOfficeRecordType.CURRENT),
```

Why not use nativeEnum here?

anikdhabal commented on packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -20,30 +22,50 @@ import {
+    value: number;
+    label: string;
+  }[] =
+    listMembers
+      ?.filter((member) => (oooType === "mine" ? me?.data?.id !== member.id : oooType === "team"))
```
I suggest using an enum instead of hardcoding

vijayraghav-io replied:
updated

anikdhabal commented on packages/features/settings/outOfOffice/OutOfOfficeEntriesList.tsx:
```diff
@@ -1,194 +1,356 @@
+            />
+            <div className="">
+              <div
+                data-testid={`member-${username}-username`}
+                className="text-emphasis text-sm font-medium leading-none">
+                {username || "No username"}
```
Refer to our team member view to see how we display the name instead of the username

vijayraghav-io replied:
updated

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.schema.ts:
```diff
@@ -2,6 +2,7 @@ import { z } from "zod";
 
 export const ZOutOfOfficeInputSchema = z.object({
   uuid: z.string().nullish(),
+  forUserId: z.number().nullable().optional(),
```
NIT:
```suggestion
  forUserId: z.number().nullish(),
```

vijayraghav-io replied:
updated

anikdhabal commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -1,26 +1,83 @@
+  const searchParams = useCompatSearchParams();
+  const router = useRouter();
+  const pathname = usePathname();
+
+  // Get a new searchParams string by merging the current searchParams with a provided key/value pair
```
Hey, always add a comment if it is very important or a complex thing to understand. I found some comments you added that are understandable. Pls remove

vijayraghav-io replied:
removed

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.test.ts:
```diff
@@ -0,0 +1,1396 @@
+      expect(ooo[0].end).toEqual("2024-10-01T23:59:59.999Z");
+    });
+
+    it("should successfully create outofoffice with redirect for logged-in user.", async () => {
+      const teams = [
+        {
+          id: 1,
+          name: "Team 1",
```
The test is more like a unit test than an integration test. We need an integration test for this type of complex handler change. If you really want this test, then move it to a different file location. Also, the test is too big. I suggested before that you only test the current changes you made because we already have end-to-end tests that cover the functionality so far. There's no point in testing similar functionality using two types of tests

vijayraghav-io commented on packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -84,6 +107,7 @@ export const CreateOrEditOutOfOfficeEntryModal = ({
+          forUserId: null,
         },
   });
   console.log(isSubmitting);
```
this was there before, not introduced from this PR, yeah it should not be there.

vijayraghav-io committed:
Update packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx

Co-authored-by: Amit Sharma <74371312+Amit91848@users.noreply.github.com>

vijayraghav-io committed:
Update packages/trpc/server/routers/loggedInViewer/outOfOffice.schema.ts

Co-authored-by: Amit Sharma <74371312+Amit91848@users.noreply.github.com>

vijayraghav-io commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -1,26 +1,83 @@
+  const isOrgAdminOrOwner =
+    data && (data.user.role === MembershipRole.OWNER || data.user.role === MembershipRole.ADMIN);
+  const toggleGroupOptions = [{ value: OutOfOfficeTab.MINE, label: t("my_ooo") }];
+  if (isOrgAdminOrOwner) {
```
I guess we have to check for Paid TeamPlan, for non-org users

vijayraghav-io replied:
updated to show for admin/owner (non-org) also

vijayraghav-io committed:
refactored to split ooo handlers, review comments

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented on packages/features/settings/outOfOffice/OutOfOfficeEntriesList.tsx:
```diff
@@ -1,194 +1,356 @@
+              <div
+                data-testid={`member-${username}-username`}
+                className="text-emphasis text-sm font-medium leading-none">
+                {username || "No username"}
```
updated as per members view

vijayraghav-io committed:
updated test.utils

vijayraghav-io commented:
@Amit91848 @anikdhabal addressed the comments

vijayraghav-io committed:
chore

vijayraghav-io committed:
chore: unrelated changes may be merge

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.utils.ts:
```diff
@@ -0,0 +1,41 @@
+import prisma from "@calcom/prisma";
+import { MembershipRole } from "@calcom/prisma/enums";
+
+export const isAdminForUser = async (adminUserId: number, memberUserId: number) => {
+  const user = await prisma.user.findUnique({
+    where: {
+      id: memberUserId,
+      teams: {
+        some: {
+          team: {
+            AND: [
+              {
+                members: {
+                  some: {
+                    userId: adminUserId,
+                    role: {
+                      in: [MembershipRole.ADMIN, MembershipRole.OWNER],
+                    },
+                    accepted: true,
+                  },
+                },
+              },
+              {
+                members: {
+                  some: {
+                    userId: memberUserId,
+                    accepted: true,
```
This is very expensive. We already have a utility function to check if a user is an admin or owner

vijayraghav-io replied:
I don't think this will be expensive, query is done on User (master) table using primary key. 
Can you please point me to the utility function, i could not find one that does the same (check input user is admin for the other user).

keithwillcode replied:
Can you please provide either the execution plan or the raw SQL this produces? We need to make sure this won’t cause perf problems before merging.

vijayraghav-io replied:
For some reason i interpreted `teams` column in `User` table as linking to `Team` table, but actually it is linked to 'Membership` table. May be i overlooked and went by the column name while building the query.
https://github.com/calcom/cal.com/blob/5bbccad41d625250ac03a6683e751b8ed6650721/packages/prisma/schema.prisma#L270

With this (knowing User table referencing to Membership table directly), i optimized the query further and came up with below query:

```
const user = await prisma.user.findUnique({
     where: {
      id: adminUserId,
      teams: {
        some: {
          accepted: true,
          role: {
            in: [MembershipRole.ADMIN, MembershipRole.OWNER],
          },
          team: {
            members: {
              some: {
                userId: memberUserId,
                accepted: true,
              },
            },
          },
        },
      },
    },
    select: {
      id: true,
    },
  });
```

For which prisma produces the raw SQL Query like below:
```
SELECT "public"."users"."id" 
FROM "public"."users" 
WHERE ("public"."users"."id" = $1 
AND ("public"."users"."id") IN (
    SELECT "t1"."userId" 
    FROM "public"."Membership" AS "t1" 
    LEFT JOIN "public"."Team" AS "j2" 
    ON ("j2"."id") = ("t1"."teamId") 
    WHERE ("t1"."accepted" = $2 
    AND (("j2"."id") IN (
        SELECT "t3"."teamId" 
        FROM "public"."Membership" AS "t3" 
        WHERE ("t3"."userId" = $3 
        AND "t3"."accepted" = $4 
        AND "t3"."role" IN (
            CAST($5::text AS "public"."MembershipRole"),
            CAST($6::text AS "public"."MembershipRole")
        ) 
        AND "t3"."teamId" IS NOT NULL
    )) 
    AND ("j2"."id" IS NOT NULL)) 
    AND "t1"."userId" IS NOT NULL
)) 
LIMIT $7 OFFSET $8;
```

Still the raw SQL Query contains join with Sub-query which is not efficient, Prisma some time produces the query with sub-queries though we intend to produce query with joins (which comparitively effecient than sub-queries).

So, Finally decided to come with 2 split simple queries that results raw SQL without joins.
```
const adminTeams = await prisma.membership.findMany({
    where: {
      userId: adminUserId,
      accepted: true,
      role: {
        in: [MembershipRole.ADMIN, MembershipRole.OWNER],
      },
    },
    select: {
      teamId: true,
    },
  });
  const adminTeamIds = adminTeams?.map((team) => team.teamId);
  const member = await prisma.membership.findFirst({
    where: {
      userId: memberUserId,
      accepted: true,
      teamId: {
        in: adminTeamIds,
      },
    },
    select: {
      id: true,
    },
  });
  ```
  The above prisma queries produces these raw SQL Queries
  
<img width="1507" alt="Screenshot 2024-09-26 at 5 57 56 PM" src="https://github.com/user-attachments/assets/39fdbc6f-288a-4572-9b7c-c52950129564">

since `userId` is indexed in `Membership` table, having these 2 simple queries, I hope should not be a performance issue.

vijayraghav-io replied:
The final raw SQL Queries generated from `isAdminForUser()` function in `outOfOffice.utilt.ts`, in text format (same as snapshot in previous comment)

```
SELECT "public"."Membership"."id", "public"."Membership"."teamId" 
FROM "public"."Membership" 
WHERE ("public"."Membership"."userId" = $1 
       AND "public"."Membership"."accepted" = $2 
       AND "public"."Membership"."role" IN (CAST($3::text AS "public"."MembershipRole"), 
                                             CAST($4::text AS "public"."MembershipRole"))) 
OFFSET $5

SELECT "public"."Membership"."id" 
FROM "public"."Membership" 
WHERE ("public"."Membership"."userId" = $1 
       AND "public"."Membership"."accepted" = $2 
       AND "public"."Membership"."teamId" IN ($3, $4)) 
LIMIT $5 OFFSET $6
```

vijayraghav-io replied:
@keithwillcode , i have attached the raw SQL queries generated from prisma query in 'isAdminForUser()' fn , can you please have look and let know if this is ok.

vijayraghav-io committed:
optimized query

vijayraghav-io commented:
@keithwillcode , reminding that have attached raw sql queries as requested, these are simple queries without complex joins, hope these should be ok, Please share if any concerns. Thankyou! 🙏

anikdhabal commented:
@vijayraghav-io thanks for your work so far. I will give the final review. Could pls fix the conflicts?

vijayraghav-io commented:
> @vijayraghav-io thanks for your work so far. I will give the final review. Could pls fix the conflicts?

sure will fix the conflicts, by tomorrow , hope it is fine.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
update due to latest changes and merge conflicts

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
@anikdhabal , Sorry for the delay!
resolved merge conflicts, and updated due to latest changes in DataTable generic component & others.
Have tested again thoroughly and its working fine.
Please take it forward.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

anikdhabal committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
updates to changes in people filter

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'feat_#16333_teamOOOCrud' of https://github.com/vijayraghav-io/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
chore

vijayraghav-io commented:
@anikdhabal , resolved merge conflicts again.
Updated to accommodate member search feature.
Tested again and verified working as expected.
Please take it to merge.

PFB loom with updated UI
https://www.loom.com/share/a20dd52901714dfaadd735f7e15893b6?sid=2229d2d9-8584-4920-9d3a-7194ac4eb2ca

vijayraghav-io committed:
as now taken care in DataTable

anikdhabal committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

Praashh reviewed (changes requested):
hey @vijayraghav-io can you resolve merge conflicts ?

vijayraghav-io commented:
> hey @vijayraghav-io can you resolve merge conflicts ?

Ok will resolve

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
@Praashh, resolved merge conflicts.

anikdhabal commented:
@vijayraghav-io Thanks for your patience, man. Could you please fix the conflicts again?

vijayraghav-io commented:
> @vijayraghav-io Thanks for your patience, man. Could you please fix the conflicts again?

sure will do

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

Praashh reviewed (dismissed):
@vijayraghav-io  unit tests are failing, please fix it 🙏

vijayraghav-io committed:
updates after merge

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
updates after merge due tp changes in datatable and others

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
update import due to file splitting and new name

vijayraghav-io commented:
> @vijayraghav-io unit tests are failing, please fix it 🙏
@Praashh resolved conflicts, updated tests due to changes from merges.
Also verified and tested again team OOO flow.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

anikdhabal committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

anikdhabal commented:
@vijayraghav-io great works man. But I am thinking of completely transferring all the E2E tests related to OOO into unit tests. Could you go through all the OOO E2E tests and add them to your current test file along with the other tests?

vijayraghav-io commented:
> @vijayraghav-io great works man. But I am thinking of completely transferring all the E2E tests related to OOO into unit tests. Could you go through all the OOO E2E tests and add them to your current test file along with the other tests?

Thank you! @anikdhabal .
IMO replacing E2E tests with unit tests may reduce the current scope of testing for OOO.
As these are End to End tests testing both UI and backend functionalities.
Also the flakiness for OOO tests is reduced very much now, observing the E2E results from latest PRs merged after https://github.com/calcom/cal.com/pull/18367

anikdhabal commented:
> IMO replacing E2E tests with unit tests may reduce the current scope of testing for OOO.
> As these are End to End tests testing both UI and backend functionalities.
> Also the flakiness for OOO tests is reduced very much now, observing the E2E results from latest PRs merged after #18367

If this is the case, then why are we adding tests to the handler here? Can't we add e2e tests instead to verify both the UI and functionalities for this PR?

I've noticed that recent PRs failed due to OOO tests. Also, I have a query regarding that PR, and we generally avoid any time dependency in e2e tests.

vijayraghav-io commented:
> I've noticed that recent PRs failed due to OOO tests. Also, I have a query regarding that PR #18367, and we generally avoid any time dependency in e2e tests.

Can you please recheck again, recent PRs don't have flakiness due to OOO tests.
Yeah E2E tests for Team OOO can be added.

vijayraghav-io commented:
These are some of E2E results after OOO flakiness fix PR (https://github.com/calcom/cal.com/pull/18367) was merged.

<img width="1097" alt="Screenshot 2024-12-29 at 9 56 24 PM" src="https://github.com/user-attachments/assets/3c0f6c80-fea8-411e-8ea5-1b98ad4acc2d" />

<img width="1167" alt="Screenshot 2024-12-29 at 10 08 56 PM" src="https://github.com/user-attachments/assets/cb4192fb-5089-44a5-8ad3-b99177ac82f0" />



This is the latest after above PR is reverted (OOO flakiness reappeared).

<img width="1167" alt="Screenshot 2024-12-29 at 10 02 18 PM" src="https://github.com/user-attachments/assets/eeb9f066-286e-4ed4-b7df-7c89886996a7" />

vijayraghav-io commented:
@anikdhabal , updated OOO E2E tests to remove flakiness without time dependency.

https://github.com/calcom/cal.com/pull/18412

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
@anikdhabal , I tried to create E2E tests for Team OOO instead of current unit tests, but currently there is some playwright limitations for this.
So, I think currently we can go with unit tests for Team OOO and once we have the solution from playwright we can create E2E tests.

The limitation is :
For Team OOO, the Create OOO Dialog is taller (because of member selection for For Member and To Member, and also made it to be minimum possible) , hence playwright can't reach the `create` button and `redirect member selection` as playwright starts with default 100% browser zoom.
It will be possible if chromium browser is zoomed out.
Currently playwright does not support browser zoom out through its APIs.
There are some issues already raised in playwright repo, refs - https://github.com/microsoft/playwright/issues/2497 , https://github.com/microsoft/playwright/issues/32352 

PFB the Loom showing the issue with a team OOO test, here i have manually zoomed out chromium browser while playwright is trying to reach down. (Here test is passing because of manul zoom out)
https://www.loom.com/share/10a0e9474d534f1b9a99b0564a8b4e98?sid=16eab3c0-6281-4de3-b0af-63443c17dc04

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
@anikdhabal , resolved merge conflicts.

Also, I tried to create E2E tests for Team OOO instead of current unit tests, but currently there is some playwright limitations for this. So, I think currently we can go with unit tests for Team OOO and once we have the solution from playwright we can create E2E tests.
The limitation is : For Team OOO, the Create OOO Dialog is taller (because of member selection for For Member and To Member, and also have made it to be minimum possible) , hence playwright can't reach the create button and redirect member selection as playwright starts with default 100% browser zoom. It will be possible if chromium browser is zoomed out. Currently playwright does not support browser zoom out through its APIs. There are some issues already raised in playwright repo, refs - [microsoft/playwright#2497](https://github.com/microsoft/playwright/issues/2497) , [microsoft/playwright#32352](https://github.com/microsoft/playwright/issues/32352)

PFB the Loom showing the issue with a team OOO test, here i have manually zoomed out chromium browser while playwright is trying to reach down. (Here test is passing because of manul zoom out during test)

https://www.loom.com/share/10a0e9474d534f1b9a99b0564a8b4e98?sid=16eab3c0-6281-4de3-b0af-63443c17dc04

Hope, with this all concerns are addressed.

anikdhabal committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'feat_#16333_teamOOOCrud' of https://github.com/vijayraghav-io/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
@anikdhabal , all tests are passing

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
update for typecheck

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
update for new add button when empty entries

vijayraghav-io commented:
resolved conflicts and updated as per new UI Components v3

vijayraghav-io committed:
nit

vijayraghav-io commented:
Tested again with latest changes

https://www.loom.com/share/a6b29d4671a541d49b12109a024d5720?sid=3e0f83d6-a3da-4727-b107-56b357383484

vijayraghav-io committed:
replace tests with e2e tests

vijayraghav-io committed:
chore

vijayraghav-io commented on apps/web/playwright/out-of-office.e2e.ts:
```diff
@@ -461,26 +462,126 @@ test.describe("Out of office", () => {
-    await dateButton.click();
-    await selectDateAndCreateOOO(page, "2", "5", "owner", 400);
-    await expect(page.locator(`text=${t("booking_redirect_infinite_not_allowed")}`)).toBeTruthy();
+  test.describe("Team OOO", () => {
+    test("Create, edit and delete", async ({ page, users }) => {
```
New tests added for team e2e

vijayraghav-io commented on apps/web/playwright/out-of-office.e2e.ts:
```diff
@@ -417,7 +418,7 @@ test.describe("Out of office", () => {
     //As owner,OOO is created on Next month 1st - 3rd, forwarding to 'member-1'
     await dateButton.click();
-    await selectDateAndCreateOOO(page, "1", "3", "member-1");
+    await selectDateAndCreateOOO(page, "1", "3", member1User?.id);
```
since team ooo dialog will have two select options (one forUserId and redirectUserId), we can't go with hardcoded member name, we have to get  member id and frame unique `data-testid`

anikdhabal commented:
@vijayraghav-io let's fix the new conflicts

vijayraghav-io commented:
> @vijayraghav-io let's fix the new conflicts

sure, will do

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
> @vijayraghav-io let's fix the new conflicts

@anikdhabal , resolved conflicts

retrogtx commented:
hey @vijayraghav-io upon adding a new ooo for a member, I am not able to click the dropdown options in the member list

![image](https://github.com/user-attachments/assets/f411346e-be10-4fb0-867d-78f4fbc67af0)

I need to type it and press enter for it to register itself
This should happen through clicking on the member name too

vijayraghav-io commented:
> hey @vijayraghav-io upon adding a new ooo for a member, I am not able to click the dropdown options in the member list
> ...

@retrogtx , can you please check again, its working as expected, able to select by clicking the option.
Please refer the video in prev comments or attached here again (showing only selection).
Also the e2e tests select the member from list without typing (or searching).

https://www.loom.com/share/94985bd76cdb4ee787a7bc4142aa00d8?sid=ce6288f2-8efb-4c0f-a6e2-56cd882e1707

vijayraghav-io commented:
also tested with firefox browser

retrogtx commented:
> also tested with firefox browser

looks like the issue is on my side, using brave browser. apologies.

retrogtx commented:
> > hey @vijayraghav-io upon adding a new ooo for a member, I am not able to click the dropdown options in the member list
> > ...
> 
> @retrogtx , can you please check again, its working as expected, able to select by clicking the option. Please refer the video in prev comments or attached here again (showing only selection). Also the e2e tests select the member from list without typing (or searching).
> 
> https://www.loom.com/share/94985bd76cdb4ee787a7bc4142aa00d8?sid=ce6288f2-8efb-4c0f-a6e2-56cd882e1707
> 
> Kindly let me know, if these are the steps you are trying to recreate or am i missing something.

It's all good, what I expected was that the drop-down should collapse upon clicking the preferred option but it seems that it has been this way and is not related to your PR. 

Thank you for confirming 🙏

vijayraghav-io commented:
> > also tested with firefox browser
> 
> looks like the issue is on my side, using brave browser. apologies.

no issues... Thanks for confirming! 🙏

anikdhabal reviewed (dismissed):
Hey @vijayraghav-io first off, great work!

We've decided to ship this feature in parts (Manage team ooo in a single pr, fetch past ooo in a single and others) instead of everything in a single PR. It's hard to review everything at once, and due to the large number of file changes, conflicts happen frequently🙏

vijayraghav-io committed:
include only manage team ooo

vijayraghav-io commented:
@anikdhabal , included only manage team ooo in this PR,
will create separate PR for fetching past records.

vijayraghav-io commented:
> @vijayraghav-io Last dance to fix the conflict🙏

ok, sure will do

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
fix type check

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/me.handler.ts:
```diff
@@ -94,6 +94,18 @@ export const meHandler = async ({ ctx, input }: MeOptions) => {
         organizationSettings: user?.profile?.organization?.organizationSettings,
       };
 
+  const isTeamAdminOrOwner =
+    (await prisma.membership.findFirst({
+      where: {
+        userId: user.id,
+        accepted: true,
+        role: { in: [MembershipRole.ADMIN, MembershipRole.OWNER] },
+      },
+      select: {
+        id: true,
+      },
+    })) !== null;
+
```
Thanks man

anikdhabal committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

anikdhabal committed:
Update safe-parse.ts

anikdhabal committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

TusharBhatt1 reviewed (commented):
One test is failing

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io committed:
use invalidate instead of props, leaving main server components as it is

vijayraghav-io commented:
> One test is failing , plus if the comments are fixed , can you just resolve them

All tests are passing now,
merged the latest main,
made one change - using invalidate query instead of having props passed across the components and also to avoid client component wrapping server component, 

Tested with latest changes again and verified team OOO working as expected. (PFB loom)
Created separated PR for fetching previous records,

https://www.loom.com/share/560b3d6ae1bf437fb71982795d370241?sid=05954ac2-6a29-4437-acf8-50d1dd732157

vijayraghav-io committed:
nit

TusharBhatt1 committed:
Merge branch 'main' into feat_#16333_teamOOOCrud

TusharBhatt1 commented on packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -156,9 +202,84 @@ export const CreateOrEditOutOfOfficeEntryModal = ({
+            {/* In case of Team, Select Member for whom OOO is created */}
+            {oooType === "team" && (
+              <>
+                <div className="mb-4">
```
We should ideally use `DropDown` here , the UI isn't ideal because of this

vijayraghav-io replied:
While editing a existing OOO record, usually we want to change dates or redirect member or reason or notes.
If we allow even editing the for member , its same as creating a new OOO, that can be done by using `Add` button, and current record can be deleted.

TusharBhatt1 replied:
My point is the UI looks weird because of it.
It remains open , even when clicked
![image](https://github.com/user-attachments/assets/92c87d57-86fd-4cb8-98f1-cd67d210e1e7)

For Edit , we can use Input (which will be disabled) , simple
Or am I missing something

vijayraghav-io replied:
> We should ideally use `DropDown` here (on Edit as well) , the UI isn't ideal because of this

yes in edit mode, the input is disabled
<img width="421" alt="Screenshot 2025-02-26 at 4 16 24 PM" src="https://github.com/user-attachments/assets/63c80684-c3db-4467-b17a-04971b9eabcc" />

TusharBhatt1 replied:
True , but what's the point of showing long div , a disabled input with team name was ideal

vijayraghav-io replied:
yeah, will update the div height

TusharBhatt1 replied:
Cool , and then for now , let's maintain the consistency and pick other enhancements in future.

vijayraghav-io replied:
<img width="422" alt="Screenshot 2025-02-26 at 4 47 44 PM" src="https://github.com/user-attachments/assets/8b597390-14a4-4d03-9f8a-c1622a278761" />

TusharBhatt1 commented on packages/features/settings/outOfOffice/CreateOrEditOutOfOfficeModal.tsx:
```diff
@@ -247,45 +368,47 @@ export const CreateOrEditOutOfOfficeEntryModal = ({
                     <Input
```
DropDown is ideal

vijayraghav-io replied:
Pls refer this comment https://github.com/calcom/cal.com/pull/16456#issuecomment-2667591691
probably for perf to fetch from large list of members and able to search, its this way

vijayraghav-io committed:
update forMember div ht in edit mode

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_#16333_teamOOOCrud

vijayraghav-io commented:
fetches 10 records at a time, when user scrolls

https://www.loom.com/share/ec62941efacc4535af92252cdcc3641e?sid=783cec35-50bd-40f9-81ab-c549b260d5d6
